lib/IRGen/GenStruct.cpp

```diff
@@ -367,6 +390,54 @@ namespace {
       (void)ClangDecl;
     }
 
+    void destroy(IRGenFunction &IGF, Address address, SILType T,
+                 bool isOutlined) const override {
+      auto *destructor = getCXXDestructor(T);
+      // If the destructor is trivial, clang will assert when we call
+      // `emitCXXDestructorCall` so, just let Swift handle this destructor.
+      if (!destructor || destructor->isTrivial()) {
+        // If we didn't find a destructor to call, bail out to the parent
+        // implementation.
+        StructTypeInfoBase<AddressOnlyClangRecordTypeInfo, FixedTypeInfo,
+                           ClangFieldInfo>::destroy(IGF, address, T,
+                                                    isOutlined);
+        return;
+      }
+
+      if (!destructor->isUserProvided()) {
+        assert(!destructor->isDeleted() &&
+               "Swift cannot handle a type with no known destructor.");
+        // Make sure we define the destructor so we have something to call.
+        auto &sema = IGF.IGM.Context.getClangModuleLoader()->getClangSema();
+        sema.DefineImplicitDestructor(clang::SourceLocation(), destructor);
+      }
+
+      clang::GlobalDecl destructorGlobalDecl(destructor, clang::Dtor_Complete);
+      auto *destructorFnAddr = cast<llvm::Function>(
+          IGF.IGM
+              .getAddrOfClangGlobalDecl(destructorGlobalDecl, NotForDefinition));
+
+      SmallVector<llvm::Value *, 2> args;
+      auto *thisArg = IGF.coerceValue(address.getAddress(),
+                                      destructorFnAddr->getArg(0)->getType(),
+                                      IGF.IGM.DataLayout);
+      args.push_back(thisArg);
+      llvm::Value *implicitParam =
+          clang::CodeGen::getCXXDestructorImplicitParam(
+              IGF.IGM.getClangCGM(), IGF.Builder.GetInsertBlock(),
+              IGF.Builder.GetInsertPoint(), destructor, clang::Dtor_Complete,
+              false, false);
+      if (implicitParam) {
+        implicitParam = IGF.coerceValue(implicitParam,
```

Comment:
Is any of the tests covering this case of having an implicit param?

Reply:
Hmm, you're right it might not be needed. The implicit param would be a vtable pointer for a base class which I think could only happen if we inherited from a virtual class in Swift. 

@martinboehme was the one who originally suggested that there was a need for this. I couldn't come up with an example where we'd need one today (maybe in the future there'd be a need) but maybe they have one in mind.

Reply:
Maybe @gribozavr remembers Martin's motivation?

Reply:
Martin wrote me in an email:

> Yes, none of the ABIs I'm aware of will actually add an implicit param in the cases that we can trigger from Swift today. Still, that's just relying on the way we happen to be using the ABIs that happen to be supported today. It would seem cleaner to keep this logic to add the implicit param if needed, though it's somewhat unsatisfactory that there is no way to test this logic today.

Reply:
I don't think it's too big of a deal; I could go either way. It's about 10 lines of code, so either I can keep it or put a comment, so future me knows what to do.